### PR TITLE
fix: document query start after

### PR DESCRIPTION
**Pull request by QuantumExplorer**
## Issue being fixed or feature implemented
Fixing issue [#2303](https://github.com/dashpay/platform/issues/2303)

## What was done?
When using an index that had subfields we needed to use conditional subqueries on the corresponding values of the start at/after document.

Consider a hierarchical index structure representing car models:

Screenshot 2024-11-25 at 04 34 40

In this example, suppose we are querying for SUVs starting after the document 29f8e... under the Escape car type of Ford. The areas marked with * denote where conditional subqueries are needed. We need to apply conditional subqueries at each level (brand_name, car_type) to ensure that the query correctly starts after the specified document.

By applying conditional subqueries:
	•	At the brand_name level: We include Ford and apply a condition to start after the specific car_type and document.
	•	At the car_type level: We include Escape and apply a condition to start after the document 29f8e....
	•	At deeper levels: The query continues without the start at condition, as it has been applied at the necessary levels.

This ensures that the query retrieves all SUVs of brand Ford and car type Escape starting after the specified document, accurately reflecting the user’s request.

## How Has This Been Tested?
Added unit tests.

## Breaking Changes
This should not be breaking.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

## Summary by CodeRabbit

- **New Features**
	- Introduced a new `StartAtDocument` struct for improved query definitions based on specific documents.
	- Added `Withdrawal` struct to manage withdrawal transactions with relevant fields.
	- Implemented new functions for generating random withdrawals and adding them to contracts.

- **Bug Fixes**
	- Updated test functions to ensure accurate querying of withdrawals based on owner ID.

- **Documentation**
	- New JSON schema added for withdrawal document structure and validation.

- **Refactor**
	- Streamlined function signatures in query logic for clarity and maintainability.

### CODE REVIEW

File: packages/rs-drive/tests/supporting_files/contract/withdrawals/withdrawals-contract.json [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+          "minimum": 1000,
+          "position": 2
+        },
+        "coreFeePerByte": {
+          "type": "integer",
+          "description": "This is the fee that you are willing to spend for this transaction in Duffs/Byte",
+          "minimum": 1,
+          "maximum": 4294967295,
+          "position": 3
+        },
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive/tests/supporting_files/contract/withdrawals/withdrawals-contract.json:98]: it's okay, just a test contract.

File: packages/rs-drive/tests/query_tests.rs [hunk 2]
```diff
@@ -434,12 +435,36 @@ struct Domain {
...
+    pub updated_at: TimestampMillis,
+
+    pub transaction_index: Option<u32>, // Optional sequential index of the transaction
+    pub transaction_sign_height: Option<u32>, // Optional Core height on which the transaction was signed
+    pub amount: u64,                          // Amount to withdraw (minimum: 1000)
+    pub core_fee_per_byte: u32,               // Fee in Duffs/Byte (minimum: 1, max: 4294967295)
+    pub pooling: u8,                          // Pooling level (enum: 0, 1, 2)
+    pub output_script: Vec<u8>,               // Byte array (size: 23-25)
+    pub status: u8,                           // Status (enum: 0 - Pending, 1 - Signed, etc.)
+}
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive/tests/query_tests.rs:460]: It's okay, just used in a test.

File: packages/rs-drive/tests/query_tests.rs [hunk 3]
```diff
@@ -599,17 +642,87 @@
...
+                } else {
+                    Identifier::random_with_rng(&mut rng)
+                };
+
+                // Determine the status randomly
+                let status = if rng.gen_bool(0.5) {
+                    0
+                } else {
+                    rng.gen_range(1..=4)
+                }; // 0 = Pending, 1-4 = other statuses
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive/tests/query_tests.rs:678]: it's okay, just used in a test.

**Review by shumkov — APPROVED** [review]
💪

**Pull request merged**